### PR TITLE
JWS: Registration Service API is authenticated with JWS

### PR DESCRIPTION
**Pull request by cpeeyush**
## What this PR changes/adds

- Access to the Registration Service API is authenticated with JWS, using scheme and libraries provided in EDC (those libraries are used in EDC for IDS authentication).
- Deploys the Registration Service controller to a custom Jersey context (port mapping), rather than the EDC default context.

See ADRs for more details.

## Why it does that

Authentication ensures that only actors who can assert their control of a did:web identifier can register that identifier as a dataspace participant.

## Linked Issue(s)

Related to https://github.com/agera-edc/MinimumViableDataspace/issues/14

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/RegistrationService/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java [hunk 1]
```diff
@@ -14,25 +14,45 @@
...
+import java.util.Objects;
+
+import static java.lang.String.format;
+
 /**
  * EDC extension to boot the services used by the Registration Service.
  */
 public class RegistrationServiceExtension implements ServiceExtension {
 
+    public static final String CONTEXT_ALIAS = "authority";
```
Comment (paullatzelsperger) [inline extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:45]: would `registration` be more suitable?

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java [hunk 2]
```diff
@@ -49,12 +69,17 @@ public class RegistrationServiceExtension implements ServiceExtension {
...
+        var authenticationService = new DidJwtAuthenticationFilter(monitor, didPublicKeyResolver, audience);
 
         participantManager = new ParticipantManager(monitor, participantStore, credentialsVerifier, executorInstrumentation);
 
         var registrationService = new RegistrationService(monitor, participantStore);
-        webService.registerResource(new RegistrationApiController(registrationService));
+        webService.registerResource(CONTEXT_ALIAS, new RegistrationApiController(registrationService));
+
+        webService.registerResource(CONTEXT_ALIAS, authenticationService);
+        webService.registerResource(CONTEXT_ALIAS, new EdcApiExceptionMapper(true));
```
[the remark this answers is not in the document]
Comment (cpeeyush) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java]: Moved to config. Thanks.

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java [hunk 3]
```diff
@@ -49,12 +69,17 @@ public class RegistrationServiceExtension implements ServiceExtension {
 
     @Override
     public void initialize(ServiceExtensionContext context) {
-        var monitor = context.getMonitor();
+        var audience = Objects.requireNonNull(context.getSetting(JWT_AUDIENCE_SETTING, null),
+                () -> format("Missing setting %s", JWT_AUDIENCE_SETTING));
+        var authenticationService = new DidJwtAuthenticationFilter(monitor, didPublicKeyResolver, audience);
```
[the remark this answers is not in the document]
Comment (algattik) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:77]: @paullatzelsperger @cpeeyush extending the RequestFilter is required in order to inject a header in the request (issuer).

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java (hunk 1 above)
Comment (algattik) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:45]: The API exposes more than just registration operations (it also allows enumerating dataspace participants).

Comment (paullatzelsperger) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:45]: But the endpoint for enumerating participants is currently available at `/authority/registry`, so maybe we could shorten this to _just_ `/registry`?

On the other hand, if we leave `/authority`, we could have one single context path for all endpoints, even future ones. In that case I would rename the extension to something like `AuthorityExtension` or similar.

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:77]: @algattik That sounds like the `DidJwtAuthenticationFilter` actually does two things:
 1. verify the signature (= "authenticate")
 2. add a header (i.e. the caller's DID)

In the spirit of "do one thing...": could we indeed convert the filter to an authentication service and use a [`RewriteHandler`](https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/rewrite/handler/RewriteHandler.html) to do the injecting of the DID? We'd have to add some glue code to the `JettyService`.|

As another option we could pass the entire request to the authentication service and have it modify the headers.

[the remark this answers is not in the document]
Comment (cpeeyush) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:77]: @paullatzelsperger One more possible idea is that EDC-AuthenticationService may return a List of AuthenticatedPrincipal or Principal and then EDC-AuthenticationRequestFilter can use them to add further information to into headers.

But this will require change in EDC code, so will that be okay if we do this separately in the next PR and merge this one as is for now ? If we agree then I will open a separate github issue on EDC repo to cover this in an upcoming PR.

[the remark this answers is not in the document]
Comment (algattik) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:77]: That's what I mean by leaky abstraction. I don't see much point in creating an overlay for the ContainerRequestFilter that has already been carefully designed in JAX-RS as a standard, especially if we are going to use JAX-RS objects in the interface...

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java (hunk 1 above)
Comment (cpeeyush) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:45]: Updated. Thanks.

File: extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (paullatzelsperger) [reply extensions/registration-service/src/main/java/org/eclipse/dataspaceconnector/registration/RegistrationServiceExtension.java:77]: The advantage of having _one_ requst filter, and _several_ `AuthenticationService` instances is to make use of the DI mechanism. Implementors of a particular authentication method need only have an `@Provider` method in their extension.

But for the time being I'm good with leaving the `DidAuthenticationFilter` as-is and merge it.

**Review by paullatzelsperger — APPROVED** [review]
